modules/heat_conduction/src/bcs/GrayLambertNeumannBC.C

```diff
@@ -0,0 +1,46 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "GrayLambertNeumannBC.h"
+#include "MathUtils.h"
+
+registerMooseObject("HeatConductionApp", GrayLambertNeumannBC);
+
+Real GrayLambertNeumannBC::_sigma_stefan_boltzmann = 5.670367e-8;
+
+template <>
+InputParameters
+validParams<GrayLambertNeumannBC>()
+{
+  InputParameters params = validParams<IntegratedBC>();
+  params.addRequiredParam<UserObjectName>("surface_radiation_object_name",
+                                          "Name of the GrayLambertSurfaceRadiation UO");
+  params.addClassDescription("This BC imposes a heat flux density that is computed from the "
+                             "GrayLambertSurfaceRadiation userobject.");
+  return params;
+}
+
+GrayLambertNeumannBC::GrayLambertNeumannBC(const InputParameters & parameters)
+  : IntegratedBC(parameters),
+    _glsr_uo(getUserObject<GrayLambertSurfaceRadiation>("surface_radiation_object_name"))
+{
+}
+
+Real
+GrayLambertNeumannBC::computeQpResidual()
+{
+  return _test[_i][_qp] * _glsr_uo.getSurfaceHeatFluxDensity(_current_boundary_id);
+}
+
+Real
+GrayLambertNeumannBC::computeQpJacobian()
+{
+  // this is not the exact Jacobian but it ensures correct scaling
+  return _test[_i][_qp] * _sigma_stefan_boltzmann * 4 * MathUtils::pow(_u[_qp], 3) * _phi[_j][_qp];
```

Comment:
Should emissivity get into here? I guess it is possible to augment the sparsity pattern for coupling of side sets and provide a function like `computeSideSetJacobian` here. But it might be an overkill and cause troubles on using the more complicated preconditioning matrix. If we get the diagonal part right, it should be a decent  Jacobian for preconditioning. I guess  we can use FD to check if assembling the non-local coupling is necessary.